### PR TITLE
[InstCombine] Fold ((X << nuw Z) binop nuw Y) >>u Z --> X binop nuw (Y >>u Z)

### PR DESCRIPTION
**Pull request by AZero13**
Proofs:
https://alive2.llvm.org/ce/z/N9dRzP
https://alive2.llvm.org/ce/z/Xrpc-Y
https://alive2.llvm.org/ce/z/BagBM6
https://alive2.llvm.org/ce/z/k2dtYZ

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1272]: do you mean `) nuw >> u Z`?

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1272]: Yeah

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1272]: Also do you mean `((X << Z) + Y) >> Z`?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1276]: This exact is only valid if `I` is `exact`

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1276]: I know. Thats' why I.isExact is there

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1276]: Oh i misread. Sorry.

Can you please update the proof to include the non-exact version and non-nsw version.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1276]: Sure: https://alive2.llvm.org/ce/z/wORNYm

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp]: No.

**Comment by AtariDreams** [comment]
@nikic What do you think about this?

**Comment by AtariDreams** [comment]
@arsenm What do you think about this PR?

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/lshr.ll:369]: Use named values in tests

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/lshr.ll:374]: Indentation off

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/lshr.ll:397]: Extra space before ( in all of these

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/lshr.ll:420]: Missing negative tests for multiple uses and individual flag missing cases

**Comment by AtariDreams** [comment]
@arsenm This is ready for review and merge! (I don't have commit perms)

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1465]: Is there a more general form for this than these specific constants? Why is this the one special case?

Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1465]: Can you add an alive2 link?

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1465]: > Is there a more general form for this than these specific constants? Why is this the one special case? 

(X+(x<<1))>>1 is canonicalized to x*3/2 so this is missed as a result

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1465]: > Can you add an alive2 link? 

Pretty sure I did in the first comment I made

**Comment by AtariDreams** [comment]
@arsenm I removed the constant specific stuff and moved a generalization of that to another PR.

**Comment by AtariDreams** [comment]
@dtcxzyw would you be so kind as to benchmark this

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/lshr.ll:363]: Please rename this test to `shl_add_lshr_flag_preservation`, and add the following basic test:
```
define i32 @shl_add_lshr(i32 %x, i32 %c, i32 %y) {
   %shl = shl nuw i32 %x, %c
   %add = add nuw i32 %shl, %y
   %lshr = lshr i32 %add, %c
   ret i32 %lshr
}
```

**Comment by AtariDreams** [comment]
@dtcxzyw I addressed all of your concerns!

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/lshr.ll:414]: Add a vector test

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/lshr.ll:374]: Extra newline

**Review by arsenm — APPROVED** [review]
test nit

**Comment by AtariDreams** [comment]
@arsenm Done!

**Review by arsenm — APPROVED** [review]

**Review by nikic — COMMENTED** [review]
This looks fine, but can you please clean up the alive2 proofs in the PR description? They seem to contain a whole bunch of stuff that is not related to the fold implemented here, like things with ashr and muls.

The proofs should all have the base shl+add+lshr sequence, just with different flags.

**Comment by AtariDreams** [comment]
@nikic @arsenm I do not have commit perms. What do you think about merging this?

**Comment by AtariDreams** [comment]
@nikic @arsenm @topperc I do not have commit perms. What do you think about merging this?

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/lshr.ll:408]: This test does not actually test  the commuted case, because it gets canonicalized, see https://llvm.godbolt.org/z/93sGo78W8.

Read https://llvm.org/docs/InstCombineContributorGuide.html#add-commuted-tests to understand how to resolve this issue.

Also, your pre-commit tests don't show this because your second commit modifies the test inputs as well. The split is to add all tests in the first commit and only change CHECK lines in the second commit. Do not modify test inputs in the second commit.

**Comment by AtariDreams** [comment]
@nikic I addressed the issues!

**Comment by AtariDreams** [comment]
@dtcxzyw can you plz run the benchmark again I wanna see something

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1297]: You can avoid the canFold variable: 

if (auto *OBO = dyn_cast<OverflowingBinaryOperator>(Op0); 
    !OBO || OBO->hasNoUnsignedWrap()) {

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1263]: This part didn't get a pattern comment

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1264]: Fixed! @arsenm

**Comment by AtariDreams** [comment]
I just realized I forgot to fix the order of the sub because the order matters, we do not canonicalize it.

I fixed the miscompile, I apologize @dtcxzyw

**Comment by AtariDreams** [comment]
Okay it is resolved now, and I added new tests @dtcxzyw, I apologize for the inconvenience.

**Comment by AtariDreams** [comment]
@dtcxzyw Though, it was thanks to the opt-benchmark that I reviewed by changes because something was off about the diff, and it did not seem equivalent at all.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1263]:
```suggestion
```
It is the same as the above line.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1268]:
```suggestion
    Value *NewLshr = Builder.CreateLShr(Y, Op1, "", /*isExact=*/true);
```

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1302]:
```suggestion
        } else if (auto *Disjoint = dyn_cast<PossiblyDisjointInst>(Op0)) {
          cast<PossiblyDisjointInst>(NewBinOp)->setIsDisjoint(Disjoint->isDisjoint());
```

**Comment by dtcxzyw** [comment]
@AtariDreams It doesn't modify anything :( https://github.com/llvm/llvm-project/compare/c8214019d1da25bb8211107e36461f9c1e7c764a..cdabaae43b855a47611a3a02ab2882f1dcd222dd

Don't rebase your patch if there is no conflict to be resolved.
See https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1307]: Which proofs are for this case? Seems like all your proofs have the `exact` flag.
Also missing a proof for `disjoint` propagation.

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/lshr.ll:374]: Missing test for `add` without nuw flag.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

Please don't force-push in this PR. The notifications flood my mailbox :(

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1288]: Although you can handle `And`, you just cant propagate `exact`: https://alive2.llvm.org/ce/z/JNaAvP

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/lshr.ll:374]: Added! @dtcxzyw

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1308]: Am I?

**Review by arsenm — APPROVED** [review]

**Comment by AtariDreams** [comment]
@arsenm I do not have commit perms by the way.

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (nikic) [reply llvm/test/Transforms/InstCombine/lshr.ll:408]: This was, once again, ***not*** resolved. Your commutative tests are still incorrect.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/lshr.ll:408]: I'll write another test.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1279]: Oops, this line doesn't match my reviewed version. https://github.com/llvm/llvm-project/commit/5d61e1f6285846d4599cd77dc6a873aa13901aaf

cc @nikic @arsenm @tstellar Is there a way to revoke the approval after new changes are made to this PR?

@AtariDreams It is dangerous to force-push non-NFC changes after receiving an approval.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1279]: > cc @nikic @arsenm @tstellar Is there a way to revoke the approval after new changes are made to this PR?

You mean in general? Github has an option to "Dismiss stale pull request approvals when new commits are pushed", but I think this is bound to multiple layers of features we do not use (required pull requests with required approvals). It also doesn't work well with the common practice where the approval contains some final nits (usually NFC nits).

For a specific PR, you can also dismiss your old approval manually, but of course this only works if you notice the change.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1279]: > Oops, this line doesn't match my reviewed version. https://github.com/llvm/llvm-project/commit/5d61e1f6285846d4599cd77dc6a873aa13901aaf
> 
> 
> 
> cc @nikic @arsenm @tstellar Is there a way to revoke the approval after new changes are made to this PR?
> 
> 
> 
> @AtariDreams It is dangerous to force-push non-NFC changes after receiving an approval.
> 
> 

And was added when another maintainer told me it worked, but only if I dropped the exact flags, and I added that proof to the proofs

**Pull request merged**
